From afbcae6d0b6582463a4ceb2a0761fb94f2de4a0b Mon Sep 17 00:00:00 2001
From: shil <shil@centecnetworks.com>
Date: Wed, 29 Jun 2022 14:34:57 +0800
Subject: [PATCH] [centec] support gd25q256 flash

---
 drivers/mtd/spi-nor/spi-nor.c | 10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/drivers/mtd/spi-nor/spi-nor.c b/drivers/mtd/spi-nor/spi-nor.c
index 848c96fc..9b0fe8d4 100644
--- a/drivers/mtd/spi-nor/spi-nor.c
+++ b/drivers/mtd/spi-nor/spi-nor.c
@@ -202,6 +202,7 @@ static inline int set_4byte(struct spi_nor *nor, const struct flash_info *info,
 		need_wren = true;
 	case SNOR_MFR_MACRONIX:
 	case SNOR_MFR_WINBOND:
+	case SNOR_MFR_GIGADEVICE:
 		if (need_wren)
 			write_enable(nor);
 
@@ -210,6 +211,10 @@ static inline int set_4byte(struct spi_nor *nor, const struct flash_info *info,
 		if (need_wren)
 			write_disable(nor);
 
+		nor->read_opcode = SPINOR_OP_READ4;
+		nor->program_opcode = SPINOR_OP_PP_4B;
+		nor->erase_opcode = SPINOR_OP_SE_4B;
+
 		return status;
 	default:
 		/* Spansion style */
@@ -850,6 +855,11 @@ static const struct flash_info spi_nor_ids[] = {
 			SECT_4K | SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ |
 			SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB)
 	},
+	{
+		 "gd25q256", INFO(0xc84019, 0, 64 * 1024, 512,
+            SPI_NOR_DUAL_READ | SPI_NOR_QUAD_READ |
+            SPI_NOR_HAS_LOCK | SPI_NOR_HAS_TB)
+	},
 
 	/* Intel/Numonyx -- xxxs33b */
 	{ "160s33b",  INFO(0x898911, 0, 64 * 1024,  32, 0) },
-- 
2.17.1

